x/axelarnet/message_handler.go

```diff
@@ -12,22 +12,87 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 
 	"github.com/axelarnetwork/axelar-core/utils/events"
+	axelarnet "github.com/axelarnetwork/axelar-core/x/axelarnet/exported"
 	"github.com/axelarnetwork/axelar-core/x/axelarnet/keeper"
 	"github.com/axelarnetwork/axelar-core/x/axelarnet/types"
 	evmtypes "github.com/axelarnetwork/axelar-core/x/evm/types"
 	nexus "github.com/axelarnetwork/axelar-core/x/nexus/exported"
 	"github.com/axelarnetwork/utils/funcs"
 )
 
+// Fee is used to pay relayer for executing cross chain message
+type Fee struct {
+	Amount    string `json:"amount"`
+	Recipient string `json:"recipient"`
+}
+
 // Message is attached in ICS20 packet memo field
 type Message struct {
 	DestinationChain   string `json:"destination_chain"`
 	DestinationAddress string `json:"destination_address"`
 	Payload            []byte `json:"payload"`
 	Type               int64  `json:"type"`
+	Fee                *Fee   `json:"fee"` // Optional
+}
+
+func (f Fee) validate(ctx sdk.Context, n types.Nexus, b types.BankKeeper, token sdk.Coin, msgType nexus.MessageType) error {
+	amt, ok := sdk.NewIntFromString(f.Amount)
+	if !ok || amt.LTE(sdk.ZeroInt()) {
+		return fmt.Errorf("invalid fee amount")
+	}
+
+	afterFee := token.Amount.Sub(amt)
+	switch msgType {
+	case nexus.TypeGeneralMessage:
+		if afterFee.LT(sdk.ZeroInt()) {
+			return fmt.Errorf("fee amount is greater than transfer value")
+		}
+	case nexus.TypeGeneralMessageWithToken:
+		if afterFee.LTE(sdk.ZeroInt()) {
+			return fmt.Errorf("fee amount is greater or equal to transfer value")
+		}
+	default:
+		return fmt.Errorf("unexpected message type for fee")
+	}
+
+	axelar := funcs.MustOk(n.GetChain(ctx, axelarnet.Axelarnet.GetName()))
+	if !n.IsAssetRegistered(ctx, axelar, token.GetDenom()) {
+		return fmt.Errorf("unregistered fee denom %s", token.GetDenom())
+	}
+
+	addr, err := sdk.AccAddressFromBech32(f.Recipient)
+	if err != nil {
+		return err
+	}
+
+	if b.BlockedAddr(addr) {
+		return fmt.Errorf("fee recipient is a blocked address")
+	}
+
+	return nil
 }
 
-func validateMessage(ctx sdk.Context, ibcK keeper.IBCKeeper, n types.Nexus, ibcPath string, msg Message, token keeper.Coin) error {
+// deductFee deducts fee from packet amount, and transfers to the recipient
+func deductFee(ctx sdk.Context, b types.BankKeeper, message Message, asset *keeper.Coin, msgID string) error {
+	if message.Fee == nil {
+		return nil
+	}
+
+	amount := funcs.MustOk(sdk.NewIntFromString(message.Fee.Amount))
+
+	events.Emit(ctx, &types.FeePaid{
+		MessageID: msgID,
+		Recipient: funcs.Must(sdk.AccAddressFromBech32(message.Fee.Recipient)),
+		Fee:       sdk.NewCoin(asset.GetDenom(), amount),
+	})
+
+	// subtract fee from transfer value
+	asset.Amount = asset.Amount.Sub(amount)
+
+	return asset.Transfer(b, types.MessageSender, funcs.Must(sdk.AccAddressFromBech32(message.Fee.Recipient)))
```

Comment:
```suggestion
	return fee.Transfer(b, types.MessageSender, funcs.Must(sdk.AccAddressFromBech32(message.Fee.Recipient)))
```
logic is reversed here, we should be transferring the fee/`amount` to the fee recipient, not the transfer amount - fee amount